### DISCUSSION TITLE
Get previous state.

### DISCUSSION THREAD

**Discussion by gonnavis** [statelyai/xstate#2671] [Q&A]
Hello, how to get the previous state?

I searched all the values of `event`, `meta` and `service.state`, but can't find it.

If I need to know the previous state, must I store it in a custom variable?

**Comment by ChrisShank** [comment]
Ya you would need to store it? May i ask why you need to know the previous state?

**Reply by gonnavis** [reply]
> Ya you would need to store it? May i ask why you need to know the previous state?

The `exit` action seems to run after state change too.

For example, in [this case](https://stately.ai/viz/3986ab65-1893-482f-86ca-1736d35a60a9), I want to do something base on the state `a` ( string 'a' ) in `exitA`. 

But in `meta`, I get state `b`, so I need to know the previous state ( state `a`, string 'a' ).

**Comment by davidkpiano** [comment]
You can grab it from `state.history`:

```js
service.subscribe(state => {
  console.log(state.history);
});
```
